restart_sampling.py

```diff
@@ -278,81 +280,153 @@ def pbar_update_absolute_wrapper(self, value, total=None, preview=None):
+                f"PlanItem: invalid restart sigmas: out of order or contains duplicates: {self}",
+            )
+            raise ValueError(errstr)
 
+    @property
+    def total_steps(self):
+        if self.k < 1:
+            return len(self.sigmas) - 1
+        return (len(self.sigmas) - 1) + (len(self.restart_sigmas) - 1) * self.k
 
-class KSamplerRestartWrapper:
-    # Some extra explanation for a couple of these arguments:
-    #
-    # chunked:
-    # When chunked is False, the sampling function is called step-by-step with only two sigmas at a time.
-    # When chunked is is True, the sampling function will be called with sigmas for multiple steps at a time.
-    # this means either the steps up to the next restart segment (or the end of sampling) or the steps within
-    # a restart segment.
-    #
-    # make_noise_sampler:
-    # If set to None, restart noise will just use torch.randn_like (gaussian) for noise generation. Otherwise
-    # this should contain a function that takes x, sigma_min, sigma_max, seed and returns a noise sampler
-    # function (which takes sigma, sigma_next) and returns a noisy tensor.
+    @property
+    def s_min(self):
+        return None if self.k < 1 else self.restart_sigmas[-1].item()
+
+    @property
+    def s_max(self):
+        return None if self.k < 1 else self.restart_sigmas[0].item()
+
+
+class RestartPlan:
     def __init__(
         self,
-        sampler,
-        real_model,
+        model,
+        steps,
+        scheduler,
+        restart_info,
         restart_scheduler,
-        restart_segments,
-        seed,
-        make_noise_sampler=None,
-        chunked=True,
+        denoise=1.0,
+        step_range=None,
+        force_full_denoise=False,
+        sigmas=None,
     ):
-        self.ksampler = sampler
-        self.real_model = real_model
-        self.restart_scheduler = restart_scheduler
-        self.restart_segments = restart_segments
-        self.total_steps = 0
-        self.seed = seed
-        self.make_noise_sampler = make_noise_sampler
-        self.chunked = chunked
+        if (
+            denoise <= 0
+            or (sigmas is None and steps < 1)
+            or (sigmas is not None and len(sigmas) < 2)
+        ):
+            self.plan = []
+            self.total_steps = 0
+            return
+        ms = model.get_model_object("model_sampling")
+
+        if sigmas is None:
+            effective_steps = steps if denoise > 0.9999 else int(steps / denoise)
+            sigmas = calc_sigmas(
+                scheduler,
+                effective_steps,
+                float(ms.sigma_min),
+                float(ms.sigma_max),
+                model.model,
+                "cpu",
+                restart_segment=False,
+            )
+        else:
+            steps = effective_steps = len(sigmas) - 1
+            steps = steps if denoise > 0.9999 else int(effective_steps * denoise)
+            sigmas = sigmas.clone().detach().cpu()
```

Comment:
yeah, i don't really know when it's necessary or not and there's weird stuff like https://github.com/blepping/comfyui_jankhidiffusion/issues/1 - you'd think doing `blah.item()` would be fine but apparently some stuff breaks without additional song and dance.

if you think `.detach()` and the no_grad stuff is unnecessary. i don't know that  `.clone()` is necessary either, i'm just scared of messing something up since i don't know 100% for sure.

Reply:
We can keep it, it's not a big deal.